plutus-core/testlib/PlutusCore/Generators/PIR/GenerateTypes.hs

```diff
@@ -0,0 +1,501 @@
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE UndecidableInstances  #-}
+{-# LANGUAGE ViewPatterns          #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+{-# OPTIONS_GHC -Wno-partial-type-signatures #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+module PlutusCore.Generators.PIR.GenerateTypes where
+
+import Control.Monad.Reader
+
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.String
+import GHC.Stack
+import Test.QuickCheck
+
+import PlutusCore.Builtin
+import PlutusCore.Default
+import PlutusCore.Generators.PIR.Common
+import PlutusCore.Name
+import PlutusCore.Normalize
+import PlutusCore.Quote (runQuoteT)
+import PlutusIR
+import PlutusIR.Error
+import PlutusIR.Subst
+
+import PlutusCore.Generators.PIR.GenTm
+
+{- Note [Debugging generators that don't generate well-typed/kinded terms/types]
+    This module implements generators for well-typed terms and well-kinded types.
+    If you came here after a property like `prop_genTypeCorrect` failed and you
+    didn't get a minimal counterexample (because shrinking requries well-typed terms)
+    you need to use a different trick. One trick that often works is to add the well-typedness
+    check in the generators - e.g. in `genTerm` you can do something like this:
+    ```
+    genTerm ... = myCheck $ do
+      ...
+      where
+        myCheck gen = do
+          (trm, type) <- gen
+          if notConformingToExpectedInvariant trm type then
+            error (show trm ++ " " ++ show type)
+          else
+            return (trm, type)
+    ```
+-}
+
+-- * Dealing with fresh names
+
+-- | Get the free variables in a type that appear in negative position
+negativeVars :: Type TyName DefaultUni () -> Set TyName
+negativeVars ty = case ty of
+  TyFun _ a b      -> positiveVars a <> negativeVars b
+  TyApp _ a b      -> negativeVars a <> negativeVars b
+  TyLam _ x _ b    -> Set.delete x $ negativeVars b
+  TyForall _ x _ b -> Set.delete x $ negativeVars b
+  TyVar _ _        -> mempty
+  TyBuiltin{}      -> mempty
+  TyIFix{}         -> error "negativeVars: TyIFix"
+
+-- | Get the free variables in a type that appear in positive position
+positiveVars :: Type TyName DefaultUni () -> Set TyName
+positiveVars ty = case ty of
+  TyFun _ a b      -> negativeVars a <> positiveVars b
+  TyApp _ a b      -> positiveVars a <> positiveVars b
+  TyLam _ x _ b    -> Set.delete x $ positiveVars b
+  TyForall _ x _ b -> Set.delete x $ positiveVars b
+  TyVar _ x        -> Set.singleton x
+  TyBuiltin{}      -> mempty
+  TyIFix{}         -> error "positiveVars: TyIFix"
+
+-- | Get the free type variables in a type along with how many
+-- times they occur. The elements of the map are guaranteed to be
+-- non-zero.
+fvTypeBag :: Type TyName DefaultUni () -> Map TyName Int
```

Comment:
This is needed for some very subtle purpose of controlling the size of the generated terms, please specify it fully in a Note.

Reply:
It is not used to generate terms. It is used to generate substitutions to test the substitution machinery however. I'm not sure it makes sense to annotate this function with documentation for how it is used, but I'll put a comment in the use site to explain what's going on there.